File: code/php/search_filters.php
<?php
require_once("xapian.php");
require_once("parsecsv.php");
require_once("logger.php");

function search($dbpath, $querystring, $materials, $offset = 0, $pagesize = 10)
{
    // offset - defines starting point within result set
    // pagesize - defines number of records to retrieve

    // Open the database we're going to search.
    $db = new XapianDatabase($dbpath);

### Start of example code.
    // Set up a QueryParser with a stemmer and suitable prefixes
    $queryparser = new XapianQueryParser();
    $queryparser->set_stemmer(new XapianStem("en"));
    $queryparser->set_stemming_strategy(XapianQueryParser::STEM_SOME);
    $queryparser->add_prefix("title", "S");
    $queryparser->add_prefix("description", "XD");

    // And parse the query
    $query = $queryparser->parse_query($querystring);

    if (empty($materials) === false)
    {
        // Filter the results to ones which contain at least one of the
        // materials.
        $material_queries = array();

        // Build a query for each material value
        foreach ($materials as $material)
        {
            $material = str_replace("material:", "", $material);
            $material_queries[] = new XapianQuery('XM'.strtolower($material));
        }

        // Combine these queries with an OR operator
        $material_query = new XapianQuery(XapianQuery::OP_AND, $material_queries);

        // Use the material query to filter the main query
        $query = new XapianQuery(XapianQuery::OP_FILTER, $query, $material_query);
    }
### End of example code.

    // Use an Enquire object on the database to run the query
    $enquire = new XapianEnquire($db);
    $enquire->set_query($query);

    // Retrieve the matches and compute start and end points
    $matches = $enquire->get_mset($offset, $pagesize);
    $match = $matches->begin();
    $end = $matches->end();

    // Use an array to record the DocIds of each match
    $docids = array();

    while (!($match->equals($end)))
    {
        // retrieve the document and its data
        $doc = $match->get_document();
        $fields = json_decode($doc->get_data());
        $position = $match->get_rank() + 1;

        // record the docid
        $docid = $match->get_docid();
        $docids[] = $docid;

        // display the results
        printf("%d: #%3.3d %s\n", $position, $docid, $fields->TITLE);

        // increment MSet iterator and our counter
        $match->next();
    }

    // Finally, make sure we log the query and displayed results
    printf("'%s'[%d:%d] = %s\n",
           $querystring,
           $offset,
           $offset+$pagesize,
           implode(" ", $docids)
    );
}

if ($argc < 2) {
    print "Usage: php $argv[0] DBPATH QUERY [MATERIALS...]\n";
    die();
}

search($argv[1], $argv[2], array_slice($argv, 3));
?>

File: code/c++/search_filters.cc
#include <xapian.h>

#include <cstdio>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

#include "support.h"

using namespace std;

static void
search(const string & dbpath, const string & querystring,
       char * materials[],
       Xapian::doccount offset = 0, Xapian::doccount pagesize = 10)
{
    // offset - defines starting point within result set.
    // pagesize - defines number of records to retrieve.

    // Open the database we're going to search.
    Xapian::Database db(dbpath);

// Start of example code.
    // Set up a QueryParser with a stemmer and suitable prefixes.
    Xapian::QueryParser queryparser;
    queryparser.set_stemmer(Xapian::Stem("en"));
    queryparser.set_stemming_strategy(queryparser.STEM_SOME);
    queryparser.add_prefix("title", "S");
    queryparser.add_prefix("description", "XD");

    // And parse the query.
    Xapian::Query query = queryparser.parse_query(querystring);

    if (materials[0] != NULL) {
	// Filter the results to ones which contain at least one of the
	// materials.

	// Build a query for each material value
	vector<Xapian::Query> material_queries;
	while (*materials) {
	    string material("XM");
	    material += Xapian::Unicode::tolower(*materials);
	    material_queries.push_back(Xapian::Query(material));
	    ++materials;
	}

	// Combine these queries with an OR operator
	Xapian::Query material_query(Xapian::Query::OP_OR,
				     material_queries.begin(),
				     material_queries.end());

	// Use the material query to filter the main query
	query = Xapian::Query(Xapian::Query::OP_FILTER, query, material_query);
    }
// End of example code.

    // Use an Enquire object on the database to run the query.
    Xapian::Enquire enquire(db);
    enquire.set_query(query);

    // And print out something about each match.
    Xapian::MSet mset = enquire.get_mset(offset, pagesize);

    clog << "'" << querystring << "'[" << offset << ":" << offset + pagesize
	 << "] =";
    for (Xapian::MSetIterator m = mset.begin(); m != mset.end(); ++m) {
	Xapian::docid did = *m;
	cout << m.get_rank() + 1 << ": #" << setfill('0') << setw(3) << did
	     << ' ';

	const size_t DOC_FIELD_TITLE = 1;
	const string & data = m.get_document().get_data();
	cout << get_field(data, DOC_FIELD_TITLE) << endl;
	// Log the document id.
	clog << ' ' << did;
    }
    clog << endl;
}

int main(int argc, char** argv) {
    if (argc < 3) {
	cerr << "Usage: " << argv[0] << " DBPATH QUERY [MATERIALS...]" << endl;
	return 1;
    }
    const char * dbpath = argv[1];

    string querystring = argv[2];

    search(dbpath, querystring, argv + 3);
}
